dvc/data/stage.py

```diff
@@ -247,27 +250,117 @@ def _get_staging(odb: "ObjectDB") -> "ObjectDB":
     return ReferenceObjectDB(fs, fs_path, state=state)
 
 
-def _load_from_state(odb, staging, fs_path, fs, name):
+def _load_raw_dir_hash(
+    odb: "ObjectDB",
+    staging: "ReferenceObjectDB",
+    fs_path: "AnyFSPath",
+    fs: "FileSystem",
+    hash_info: "HashInfo",
+    meta: "Meta",
+    name: "str",
+    dry_run: bool,
+) -> "Tree":
+    from dvc.fs._callback import NoOpCallback
+    from dvc.ui import ui
+
+    from .tree import Tree
+
+    tree = _try_load(odb, hash_info.as_raw(), raw_dir_hash=True)
+    if tree is None:
+        raise FileNotFoundError
+
+    assert isinstance(tree, Tree)
+    assert tree.fs
+    meta.nfiles = len(tree)
+    assert tree.hash_info.name == name
+
+    if not dry_run:
+        with ui.status(f"Loading {fs_path}"):
+            for key, _, oid in tree:
+                staging.add(
+                    fs.path.join(fs_path, *key),
+                    fs,
+                    oid,
+                    hardlink=False,
+                    verify=False,
+                    callback=NoOpCallback(),
+                )
+
+            staging.add(
+                tree.fs_path,
+                tree.fs,
+                hash_info,
+                hardlink=False,
+                callback=NoOpCallback(),
+            )
+
+    raw = staging.get(hash_info)
+    tree.fs = raw.fs
+    tree.fs_path = raw.fs_path
+    tree.hash_info = hash_info
+    return tree
+
+
+def _try_load(
+    odb: "ObjectDB",
+    hash_info: "HashInfo",
+    raw_dir_hash: bool = False,
```

Comment:
Does `_try_load()` need this flag (and/or is this method needed at all)?

I get that this is trying to reduce duplicated load/check code between the old `_load_from_state` and the new raw tree handling, but I think this might end up making it more confusing to read.

It seems to me like this is adding a helper for loading any generic object, but in the event that the caller knows they want to load a specific hash as a specific type of object, I think the caller should just be doing that directly (rather than passing extra flags into a helper method).

So in this case, instead of using a helper method `_load_raw_dir_hash` would just start with

```py
    try:
        tree = Tree.load(odb, hash_info.as_raw())
        tree.check(odb)
    except ObjectFormatError:
        raise FileNotFoundError
```

Alternatively, if you want to keep the helper methods to reduce the exception handling boilerplate in the event we need to re-use this functionality again later, I'd prefer to see separate generic `_try_load()` and `_try_load_tree()` helper methods.

Reply:
Basically, when we have code that looks like:

```py
tree = _try_load(odb, hash_info.as_raw(), raw_dir_hash=True)
assert isinstance(tree, Tree)
```

It leads me to think that the helper method is not actually helping, given that we are already doing 2 different things to use the "raw" hash, and we are also forced to verify that the helper is in fact returning what we expected it to return.

(Not sure if the assert statement is to clean up mypy warnings, but in this case I would agree with mypy that it is confusing to figure out what `_try_load()` is going to do if I was looking at that line in the code)